runtime/tests/tests.rs

```diff
@@ -222,35 +263,318 @@ fn generic_asset_transfer_works_with_fee_exchange() {
+				balance_amount - transfer_amount,
+			);
+			assert_eq!(
+				<GenericAsset as MultiCurrencyAccounting>::free_balance(&alice(), Some(CENNZ_ASSET_ID)),
+				balance_amount - 260_346_803_274,
+			);
+		});
 }
 
 #[test]
 fn contract_call_fails_when_fee_exchange_is_not_enough_for_gas() {
-	// Fee exchange not enough to pay for gas
-	// validate() should early terminate?
+	let contract_call = Call::Contracts(pallet_contracts::Call::call::<Runtime>(
+		bob(),
+		50,
+		10 * ContractTransactionBaseFee::get() as u64,
+		vec![],
+	));
+
+	ExtBuilder::default()
+		.initial_balance(10_000 * TransactionBaseFee::get())
+		.gas_price(1)
+		.build()
+		.execute_with(|| {
+			let liquidity_core_amount = 100 * TransactionBaseFee::get();
+			let liquidity_asset_amount = 10 * TransactionBaseFee::get();
+			let _ = CennzxSpot::add_liquidity(
+				Origin::signed(charlie()),
+				CENNZ_ASSET_ID,
+				10, // min_liquidity
+				liquidity_asset_amount,
+				liquidity_core_amount,
+			);
+			let ex_key = (CENTRAPAY_ASSET_ID, CENNZ_ASSET_ID);
+			assert_eq!(CennzxSpot::get_liquidity(&ex_key, &charlie()), liquidity_core_amount);
+
+			let fee_exchange = FeeExchange::V1(FeeExchangeV1 {
+				asset_id: CENNZ_ASSET_ID,
+				max_payment: 1,
+			});
+
+			let xt = sign(CheckedExtrinsic {
+				signed: Some((alice(), signed_extra(0, 0, None, Some(fee_exchange)))),
+				function: contract_call,
+			});
+			initialize_block();
+			let r = Executive::apply_extrinsic(xt);
+			assert_eq!(r, Err(InvalidTransaction::Payment.into()));
+		});
 }
 
 #[test]
 fn contract_call_fails_when_exchange_liquidity_is_low() {
-	// Exchange doesn’t have sufficient liquidity
+	let gas_limit_amount = 10 * ContractTransactionBaseFee::get();
+	let contract_call = Call::Contracts(pallet_contracts::Call::call::<Runtime>(
+		bob(),
+		50,
+		gas_limit_amount as u64,
+		vec![],
+	));
+
+	ExtBuilder::default()
+		.initial_balance(10_000 * TransactionBaseFee::get())
+		.gas_price(1)
+		.build()
+		.execute_with(|| {
+			let liquidity_core_amount = 100;
+			let liquidity_asset_amount = 100;
+			let _ = CennzxSpot::add_liquidity(
+				Origin::signed(charlie()),
+				CENNZ_ASSET_ID,
+				10, // min_liquidity
+				liquidity_asset_amount,
+				liquidity_core_amount,
+			);
+			let ex_key = (CENTRAPAY_ASSET_ID, CENNZ_ASSET_ID);
+			assert_eq!(CennzxSpot::get_liquidity(&ex_key, &charlie()), liquidity_core_amount);
+
+			let fee_exchange = FeeExchange::V1(FeeExchangeV1 {
+				asset_id: CENNZ_ASSET_ID,
+				max_payment: 100_000_000 * gas_limit_amount,
+			});
+
+			let xt = sign(CheckedExtrinsic {
+				signed: Some((alice(), signed_extra(0, 0, None, Some(fee_exchange)))),
+				function: contract_call,
+			});
+			initialize_block();
+			let r = Executive::apply_extrinsic(xt);
+			assert_eq!(r, Err(InvalidTransaction::Payment.into()));
+		});
+}
+
+#[test]
+fn contract_call_with_cpay_fee_exchange_does_not_use_exchange() {
```

Comment:
Might need to rename this, originally we thought we could ignore the fee_exchange, but now we know the transaction will be rejected.
Maybe something like
`contract_call_with_cpay_fee_exchange_fails`

Reply:
How about `contract_call_fails_when_cpay_is_used_for_fee_exchange`